apps/emqx_bridge_s3/src/emqx_bridge_s3_aggregator.erl

```diff
@@ -0,0 +1,486 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2022-2024 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+%% This module manages buffers for aggregating records and offloads them
+%% to separate "delivery" processes when they are full or time interval
+%% is over.
+-module(emqx_bridge_s3_aggregator).
+
+-include_lib("emqx/include/logger.hrl").
+-include_lib("snabbkaffe/include/trace.hrl").
+
+-include("emqx_bridge_s3_aggregator.hrl").
+
+-export([
+    start_link/2,
+    push_records/3,
+    tick/2,
+    take_error/1
+]).
+
+-behaviour(gen_server).
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    terminate/2
+]).
+
+-export_type([
+    record/0,
+    timestamp/0
+]).
+
+%% Record.
+-type record() :: #{binary() => _}.
+
+%% Unix timestamp, seconds since epoch.
+-type timestamp() :: _Seconds :: non_neg_integer().
+
+%%
+
+-define(VSN, 1).
+-define(SRVREF(NAME), {via, gproc, {n, l, {?MODULE, NAME}}}).
+
+%%
+
+start_link(Name, Opts) ->
+    gen_server:start_link(?SRVREF(Name), ?MODULE, mk_state(Name, Opts), []).
+
+push_records(Name, Timestamp, Records = [_ | _]) ->
+    %% FIXME: Error feedback.
+    case pick_buffer(Name, Timestamp) of
```

Comment:
Could be selected buffer just invalidated after pick?
Would that result e.g. in rotating buffer twice?

Reply:
Yes it can. The idea is that `file:write/2` then returns `{error, terminated}`, which will translate to `rotate_buffer/2`, and if the buffer in question has been already rotated by the time `handle_rotate_buffer/2` aggregator process will just respond with the current buffer (since it has different FD), which will most likely be a buffer for the next time window in this case.

Though I may be missing something important, please write down a scenario that you think could lead to weird outcomes.

Reply:
As far as I understood, on failure, `write_records` tries to obtain another buffer through `rotate_buffer` which unconditionally makes rotation...
I was thinking about a paranoid scenario when, due to concurrency, several writers invalidate each other's buffers and leave many closed empty buffers 😅

Reply:
Yeah, that indeed could happen! There's [this clause](https://github.com/emqx/emqx/blob/3afee4ed8b2ac8f557a2b7eeeae4e4a4456bee69/apps/emqx_bridge_s3/src/emqx_bridge_s3_aggregator.erl#L228) that (hopefully) should basically deduplicate those _rotate_ requests.